### PR TITLE
C++11 override and cbegin/cend fixes

### PR DESCRIPTION
**Pull request by ghost**
This adds override, cbegin, cend, a copy constructor, floating point fixes, and a few other things to various parts of GWork. As well, various variables are being shadows; this corrects that.

Without this fix, both Clang 3.8 and GCC 5.3.0 do not allow GWork to compile.

### CODE REVIEW

**Comment by billyquith** [comment]
> Without this fix, both Clang 3.8 and GCC 5.3.0 do not allow GWork to compile.

Which platform are you on? I'm compiling on Mac Clang and this works fine. There are warnings but it compiles and works. Do you have "warnings as errors" set?

File: source/gwork/include/Gwork/ControlList.h [hunk 1]
```diff
@@ -48,17 +49,17 @@ namespace Gwk
             list.remove(control);
         }
 
-        void Add(const List& list)
+        void Add(const List& _list)
```
Comment (billyquith) [inline source/gwork/include/Gwork/ControlList.h:52]: What seems to be the problem here? Are you doing `using namespace std`?

File: source/gwork/source/Anim.cpp [hunk 2]
```diff
@@ -113,9 +113,9 @@ static float GetEased(float fTime, float fEase)
 {
     // Ease in and Out if ease is < 0
     if (fEase < 0)
-        return -fTime/2*(std::cosf(3.14159f*fTime)-1);
+        return -fTime/2*(std::cos(3.14159f*fTime)-1);
```
Comment (billyquith) [inline source/gwork/source/Anim.cpp:116]: I should have used C99 here: `cosf()`

File: source/gwork/include/Gwork/Skins/TexturedBase.h [hunk 3]
```diff
@@ -1006,7 +1003,7 @@ namespace Gwk
                 Textures.Input.Slider.H.Normal.DrawCenter( GetRender(), control->GetRenderBounds() );
             }
 
-            void DrawTreeNode( Controls::Base* ctrl, bool bOpen, bool bSelected, int iLabelHeight, int iLabelWidth, int iHalfWay, int iLastBranch, bool bIsRoot )
+            virtual void DrawTreeNode( Controls::Base* ctrl, bool bOpen, bool bSelected, int iLabelHeight, int iLabelWidth, int iHalfWay, int iLastBranch, bool bIsRoot ) override
```
Comment (billyquith) [inline source/gwork/include/Gwork/Skins/TexturedBase.h:1006]: These don't need to be virtual if they are override.

File: source/designer/source/DocumentCanvas.cpp [hunk 4]
```diff
@@ -67,7 +67,7 @@ void DocumentCanvas::SelectControls( ControlList& CtrlList )
 {
 	m_selectionLayer->ClearSelection();
 
-	for ( ControlList::List::const_iterator it = CtrlList.list.begin(); it != CtrlList.list.end(); ++it )
+	for ( ControlList::List::const_iterator it = CtrlList.list.cbegin(); it != CtrlList.list.cend(); ++it )
```
Comment (billyquith) [inline source/designer/source/DocumentCanvas.cpp]: Good attention to detail. These will probably all get changed to for-range over time.

**Comment by ghost** [comment]
@billyquith Clang 3.8 on Ubuntu; I've checked on Gentoo with GCC 5.3.0 and the same errors occur; notably, std::find cannot be found without including algorithm. Mac Clang probably has some sort of automatic include like VS does.

File: source/gwork/include/Gwork/ControlList.h (hunk 1 above)
Comment (ghost) [inline source/gwork/include/Gwork/ControlList.h:52]: The local variable list shadows the class variable list; this causes clang to complain, and makes it kind of hard to reason about the code.

File: source/gwork/source/Anim.cpp (hunk 2 above)
Comment (ghost) [inline source/gwork/source/Anim.cpp:116]: C++11 cos is overloaded for float and double, so just using cos is fine; it's not guaranteed C99 cosf will be put under std:: (and it isn't on VS2015, or GCC under linux).

This is one error preventing GWork from compiling on my system

File: source/gwork/include/Gwork/Skins/TexturedBase.h (hunk 3 above)
Comment (ghost) [inline source/gwork/include/Gwork/Skins/TexturedBase.h:1006]: They don't need to be, but they will be virtual implicitly. If you like, I can remove these, if you find the lack of a virtual more self documenting or something.

File: source/designer/source/DocumentCanvas.cpp (hunk 4 above)
Comment (ghost) [inline source/designer/source/DocumentCanvas.cpp]: I was going to replace them with range for, but wanted to check if that was what you intended first. I noticed the code base was sticking to a bit more of a C++98 style, despite billing itself as C++11.

File: source/gwork/include/Gwork/Skins/TexturedBase.h (hunk 3 above)
Comment (billyquith) [inline source/gwork/include/Gwork/Skins/TexturedBase.h:1006]: The override is the important bit. The virtual here is redundant, and if anything misleading. If you only see virtual at the first parent method then you know that is the "base". When you see override you know a) it is overriding something and b) the compiler will check that the override is correct.

Override is great in a codebase like this with loads of child classes when you are refactoring heavily. It is really annoying tracking down bugs due to virtual failures. :(

If you don't mind removing the redundant virtuals (i.e. where it is an override) that would be great.

File: source/designer/source/DocumentCanvas.cpp (hunk 4 above)
Comment (billyquith) [inline source/designer/source/DocumentCanvas.cpp]: If you have a look at the GWEN code, which is pre-C++11, you'll see it has changed quite a lot already. GWEN "works" but there are quite a lot of improvements that can be made. TBH it might have been faster to start from scratch!

Changing these makes the code look nicer, but isn't essential. If you'd like to do it then that would be great.

**Comment by billyquith** [comment]
I've had similar problems with Ponder cross platform. Linux seems to require more header includes. I should probably set up Travis and Appveyor for gwork.

File: source/gwork/include/Gwork/Skins/TexturedBase.h (hunk 3 above)
Comment (ghost) [inline source/gwork/include/Gwork/Skins/TexturedBase.h:1006]: While I'd be willing to remove the redundant virtuals, I'd prefer to do so as part of a larger overhaul; TextBox.h, for example, has about a full page worth that need to be removed.

I'd be willing to go through and get rid of all of them, but it may be better to do that in a separate pull request.

Comment (billyquith) [inline source/gwork/include/Gwork/Skins/TexturedBase.h:1006]: Don't worry about it. If you add all the overrides (and clang gives warning help) then I think I can write a regex to get rid of the virtuals.

File: source/designer/source/DocumentCanvas.cpp (hunk 4 above)
Comment (ghost) [inline source/designer/source/DocumentCanvas.cpp]: I think keeping a GWEN-like interface was probably a better idea than attempting to do a full from scratch rewrite; there are already plenty of other UI libraries.

I've replaced most of the loops with range-for when it is trivial to do so.

File: source/gwork/include/Gwork/Skins/TexturedBase.h (hunk 3 above)
Comment (ghost) [inline source/gwork/include/Gwork/Skins/TexturedBase.h:1006]: I've added all overrides I think applicable.

**Comment by ghost** [comment]
All changes done; this is ready to pull, unless there are further enhancements or changes needed.

File: source/gwork/include/Gwork/Controls/CollapsibleList.h [hunk 5]
```diff
@@ -61,10 +61,10 @@ namespace Gwk
             {
                 Base::List& children = GetChildren();
 
-                for (Base::List::iterator iter = children.begin(); iter != children.end(); ++iter)
+                for (auto&& iter : children)
```
Comment (billyquith) [inline source/gwork/include/Gwork/Controls/CollapsibleList.h]: Good use of `auto&&`. I keep forgetting about this. I swap between languages a lot!

This is no longer an iterator though, you have unwrapped it by using the for-range, so it is now a child, e.g.

``` cpp
for (auto&& child : children) { ... }
```

I know this sounds pedantic, but the point of using the for-range is to make the code easier to read/understand, so the name is important.

Comment (ghost) [inline source/gwork/include/Gwork/Controls/CollapsibleList.h]: Notice that iter is immedietely assigned into a new variable, after being cast through a gwk_cast.
That variables name is.... `child.`

Everywhere applicable, I used a more descriptive variable name; this just wasn't one of those places. If you want a more descriptive variable name you'll need to provide one!

I could also possibly to normalize these variable names, since it and iter are used interchangably for some reason, but I defer to the project owner for such things. Maybe "instance"?

All good to go if you don't want me to change the variable names.

File: source/designer/source/DocumentCanvas.cpp (hunk 4 above)
Comment (ghost) [inline source/designer/source/DocumentCanvas.cpp]: Replaced more loops!

File: source/gwork/include/Gwork/Skins/TexturedBase.h (hunk 3 above)
Comment (ghost) [inline source/gwork/include/Gwork/Skins/TexturedBase.h:1006]: Added a few more!

**Comment by Alicemargatroid** [comment]
Just need some feedback on what you want me to name these iterator variables, and this should be good to go. I'd love to name it `child` but that is unfortunately taken.

File: source/gwork/include/Gwork/Controls/CollapsibleList.h (hunk 5 above)
Comment (billyquith) [inline source/gwork/include/Gwork/Controls/CollapsibleList.h]: Well, with using auto I think it is a good idea to have more descriptive names to give hints about the type. `control` or `ctrl` would do since that what the child list contains. Then that is cast to a child.

Todo: rename `Base` to `Control`. - not you!

Comment (ghost) [inline source/gwork/include/Gwork/Controls/CollapsibleList.h]: @billyquith As many range-for names changed as possible!

**Comment by ghost** [comment]
Descriptive names used when possible during range-for's; all ready to go!
Just fixed the syntax to sit more seamlessly with the prior code.

**Comment by billyquith** [comment]
Thanks. Much more readable. Using auto is useful but removes information for the reader.
